### PR TITLE
[wip] Unify checkpoint load paths

### LINKED ISSUE #9405: [checkpoint] Resolve 2 different checkpoint loading paths across `fit` vs `validate`/`test`/`predict`
## Proposed refactoring or deprecation

Consolidate checkpoint loading code across `fit` and `validate`/`test`/`predict`

### Motivation

We have 2 different code paths for checkpoint loading
1. `Trainer.fit` and the constructor argument `resume_from_checkpoint`
2. `ckpt_path` passed to `Trainer.validate/test/predict`

Offering multiple code paths here risks divergence. Lightning must ensure a consistent experience for checkpoint loading across these different entry points.

### Background
These are the paths today.

Trainer.fit:
- Trainer constructor is initialized with `resume_from_checkpoint`: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L143
- Which initializes the checkpoint connector with `resume_from_checkpoint`: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L383
- Which sets this field in the connector: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L38
- In `_run` we call `Trainer._restore_modules_and_callbacks`: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L960-L962
- Which explicitly checks if we're fitting, and if so, first loads the checkpoint, before restoring the model 

Trainer.validate/test/predict:
- `ckpt_path` is an argument to the function
- There's a common helper to validate the checkpoint path: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L1200
- A property is set on the trainer for the corresponding path: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L671-L673
- We use an umbrella property to work across validate/test/predict: https://github.com/PyTorchLightning/pytorch-lightning/blob/6e124e7207f6459cb43f540cfb5a1c6cc9b00f7a/pytorch_lightning/trainer/properties.py#L626-L633
- There are checks for loading based on if this property is set inside `Trainer._run`
-- https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L950-L951
-- https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L1006-L1007

Notably:
- trainer.fit calls `restore_model` whereas `trainer.validate/test/predict` calls `restore_model_weights` in the checkpoint connector. https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L116-L150
- Load checkpoint weights is called before the accelerator setup while restore model is called after: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L950-L951r: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L960-L962

### Pitch

1. Unify in CheckpointConnector
`CheckpointConnector.restore_model` is a superset of `CheckpointConnector.restore_model_weights` which suggests we don't need both. 

2. Unify in trainer.py

The overall sequence in `_run` can be shared:
- Load & validate the checkpoint dict through the training type plugin
- Load the datamodule state
- Load the LightningModule state through the training type plugin
- Load the callback states
- Load the loop states
- If fitting, load the precision, optimizer, and LR scheduler states

We have 5 different properties exposed for the checkpoint path to resume from (excluding HPC stuff):
- `trainer.resume_from_checkpoint`: 
- `trainer.validated_ckpt_path`: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L671-L674
- `trainer.tested_ckpt_path`: https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L760-L762
- `trainer.predicted_ckpt_path` : https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/trainer.py#L843-L845
- An umbrella property across these 3: `trainer._ckpt_path`: https://github.com/PyTorchLightning/pytorch-lightning/blob/6e124e7207f6459cb43f540cfb5a1c6cc9b00f7a/pytorch_lightning/trainer/properties.py#L626-L633
-- This is also inconsistent: the attributes initialized in validate/test/predict are public, while `_ckpt_path` is private. Why?

It's unclear what the lifecycle of these properties should be. Do successive calls to validate/test/predict end up relying on this?

Proposal: 
- Pass the checkpoint path to resume from as an argument to `Trainer._run` to avoid our dependency on these properties
- Deprecate these properties

3. Unify in the Trainer API
Proposal: deprecate `resume_from_checkpoint` from the Trainer constructor, and add a new argument `ckpt_path` to `Trainer.fit` . This provides API consistency with `validate`/`test`/`predict`

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #9405 
## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -943,24 +955,18 @@ def tune(
...
-        if not self._device_type == DeviceType.TPU:
-            self.accelerator.barrier()
-        rank_zero_info(f"Loading model weights from checkpoint at {self._ckpt_path}")
-        self.checkpoint_connector.restore_model_weights(self._ckpt_path)
-
-    def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
+        if self.state.fn == TrainerFn.FITTING:
+            self.checkpoint_connector.restore_datamodule()
+            # restore callback states
+            self.checkpoint_connector.restore_callbacks()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:965]: we should restore datamodule & callback states even if we're not fitting right?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1038,12 +1041,11 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
...
         if self.accelerator.restore_checkpoint_after_pre_dispatch:
-            if self._ckpt_path:
-                self._load_checkpoint_weights()
-            self._restore_modules_and_callbacks()
+            self._restore_modules_and_callbacks(ckpt_path)
 
         # restore optimizers, etc.
-        self.checkpoint_connector.restore_training_state()
+        if self.state.fn == TrainerFn.FITTING:
+            self.checkpoint_connector.restore_training_state()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1048]: restore training state includes things which can be resumed even if not fitting, such as the loop state.

imo we shouldn't add the check for fitting here, but rather inside the select parts inside of `restore_training_state`

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1733,7 +1735,7 @@ def checkpoint_callbacks(self) -> List[ModelCheckpoint]:
         return [c for c in self.callbacks if isinstance(c, ModelCheckpoint)]
 
     @property
-    def resume_from_checkpoint(self) -> Optional[Union[str, Path]]:
+    def resume_checkpoint_path(self) -> Optional[Union[str, Path]]:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1738]: - let's use the typehint _PATH here too? 
- should this property be deprecated as well given it's entirely user input?

**Review by ananthsub — COMMENTED** [review]
this is looking really promising @jjenniferdai ! unifying these paths will make the API consistent and help us simplify the trainer internals

fyi @carmocca who was interested in this too from #9405

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -37,7 +37,12 @@
 class CheckpointConnector:
     def __init__(self, trainer: "pl.Trainer", resume_from_checkpoint: Optional[_PATH] = None) -> None:
         self.trainer = trainer
-        self.resume_checkpoint_path = resume_from_checkpoint
+        self.resume_checkpoint_path: Optional[_PATH] = None
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:40]: Do we need to keep this attribute?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:965]: i was wondering this too. currently i think only fitting restores `restore_datamodule`, `restore_callbacks`, and `restore_training_state` as these fns all return early if `checkpoint_connector._loaded_checkpoint` is not set, `checkpoint_connector._loaded_checkpoint` is only set in `resume_start` which is only called for fitting?

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -438,11 +445,6 @@ def __init__(
         # Needed because of LightningOptimizer
         self._lightning_optimizers = None
 
-        # .validate() and .test() set this when they load a checkpoint
-        self.validated_ckpt_path: Optional[str] = None
-        self.tested_ckpt_path: Optional[str] = None
-        self.predicted_ckpt_path: Optional[str] = None
-
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:445]: detailed in issue #9405, can directly pass the path to `_run` instead of relying on these properties

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:1738]: yes on second thought i think this can be deprecated entirely. user passes`ckpt_path`, `resume_start`/`resume_end` will be called to set and forget this in checkpoint_connector

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:40]: I think so, it helps CheckpointConnector remember which path `_loaded_checkpoint` is from, instead of passing a path to every function and loading path file contents each time. wdyt?

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:1048]: (above comment)

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:451]: The intended use for these is to be able to get the actual `ckpt_path` used when `ckpt_path="best"` is passed.

Unifying the load paths shouldn't have anything to do with them.

These are ancient attributes. We can discuss removing them but IMO it should be done in a separate PR.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:40]: If we deprecate the following: https://github.com/PyTorchLightning/pytorch-lightning/pull/9693/files#r716905724
then I think we can remove it. The only thing that's not issue is keeping `rank_zero_info` message in `resume_end`

Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:40]: aside from losing the message, would it be odd keeping `_loaded_checkpoint` without `resume_checkpoint_path` (where is `_loaded_checkpoint` from?)

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:40]: I wouldn't say so

Any preference @ananthsub or @awaelchli?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:40]: I recommend not keeping the loaded checkpoint in memory beyond the resuming procedure. Please keep the log message acknowledging loading of the checkpoint and the path to it.

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:973]: @awaelchli @carmocca @rohitgr7 do you know if this was intentional?

if not fitting is it expected to only load the weights, and not data/callback states? @rohitgr7 is fixing the trainer state when not fitting here: https://github.com/PyTorchLightning/pytorch-lightning/issues/9404

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -1740,10 +1742,6 @@ def checkpoint_callbacks(self) -> List[ModelCheckpoint]:
         in the Trainer.callbacks list."""
         return [c for c in self.callbacks if isinstance(c, ModelCheckpoint)]
 
-    @property
-    def resume_from_checkpoint(self) -> Optional[Union[str, Path]]:
-        return self.checkpoint_connector.resume_checkpoint_path
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1745]: we can't delete this outright. this is a public property, part of the trainer API, which needs to go through the deprecation cycle before its removal

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:973]: in #9413 we are now loading the checkpoint to retrieve the `global_step` and `current_epoch` if not fitting and nothing else. Not sure about datamodule but for callbacks we never loaded any states if not fitting. we have an old issue to discuss whether to load the callback states or not here: https://github.com/PyTorchLightning/pytorch-lightning/issues/5542

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1056]: yeah, we do restore loops there and now some other attributes as well. I'd suggest waiting for this one to get merged: https://github.com/PyTorchLightning/pytorch-lightning/pull/9413

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -595,14 +604,17 @@ def fit(
...
+            self._fit_impl, model, train_dataloaders, val_dataloaders, datamodule, ckpt_path
+        )
 
     def _fit_impl(
         self,
         model: "pl.LightningModule",
         train_dataloaders: Optional[Union[TRAIN_DATALOADERS, LightningDataModule]] = None,
         val_dataloaders: Optional[EVAL_DATALOADERS] = None,
         datamodule: Optional[LightningDataModule] = None,
+        ckpt_path: Optional[str] = None,
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:617]: should this also be typed as `_PATH` ?

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -1046,12 +1051,13 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
...
-                self._load_checkpoint_weights()
-            self._restore_modules_and_callbacks()
+            self._restore_modules_and_callbacks(ckpt_path)
 
         # restore optimizers, etc.
-        self.checkpoint_connector.restore_training_state()
+        if self.state.fn == TrainerFn.FITTING:
+            self.checkpoint_connector.restore_training_state()
+
+        self.checkpoint_connector.resume_end()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1060]: n00b question: why is this bumped up to here vs in `_pre_training_routine` ?

Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:1060]: now that this calls `resume_start` for all (not only fitting), similarly `resume_end` for all as well

**Comment by tchaton** [comment]
Dear @@jjenniferdai,

Any updates ?

Best,
T.C

**Comment by jjenniferdai** [comment]
sorry all for the git issues :( please see #10061 instead
